interfaces/cython/cantera/composite.py

```diff
@@ -530,6 +536,19 @@ def __init__(self, phase, shape=(0,), states=None, extra=None):
                 " supplied in a dict if the SolutionArray is not initially"
                 " empty")
 
+        # add properties dynamically
+        state_sets = set(phase._full_states.values()) | set(phase._partial_states.values())
+        state_sets = state_sets | set(self._all_states)
+        for name in state_sets:
+            ph = type(phase)
+            if len(name) == 2:
+                setattr(SolutionArray, name, _state2_prop(name, ph))
+            elif len(name) == 3:
+                setattr(SolutionArray, name, _state3_prop(name, ph))
+            else:
+                raise NotImplementedError("Failed adding property '{}' for "
+                                          "phase '{}'".format(name, phase))
```

Comment:
This is indeed the point. Since the lists of setters are no longer available as static variables in python (they are now defined in the C++ layer), I didn't see an alternative. Suggestions?

Regarding the `PureFluid` issue, this is temporary and can be cleaned up once `TQ` and `TX` are clearly separated in #719.

Reply:
The concern isn't about `TX` in particular, since I realize that that will change with #719. I picked it because it's the one case I know of where we have a state tuple that's supported for a model that isn't present in the base `ThermoPhase` class.

You can avoid modifying the base class with the approach in 2550e4860cba217def5fe055425887731cde4ea4, where I've basically replace the properties with the use of `__getattr__` and `__setattr__` to provide the same interface. However, both this approach and the one you have implemented have the downfall of removing all of these properties from the documentation. This approach has the additional side effect that none of these properties show up in `dir` or the autocomplete options in IPython, which is unfortunate.

One alternative would be to continue with your approach, but add the base set of properties from `SolutionArray._all_states` when the module is loaded (i.e. the behavior before this commit) and then add any others implemented by a particular phase here. That way, the properties from `_all_states` would at least be documented.

Reply:
@speth ... I like your suggestion (I.e. the referenced commit). How would you go about the documentation for that case? I figure one way would be to add all and raise exceptions when they’re not available?

Reply:
OK. I applied the suggested solution *verbatim*. As #719 will provide a really good use case, could we defer the remainder of this discussion for the next PR?

Reply:
PS: since `SolutionArray` doesn't define these setters/getters itself, I am actually not sure that the documentation needs to be inherited from `Solution`, `PureFluid`, etc.. I.e. I believe that your suggestion may be the preferable approach anyways.

Reply:
Cherry-picked resolution. Thank you @speth